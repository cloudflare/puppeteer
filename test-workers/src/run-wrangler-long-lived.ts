import {fork} from 'node:child_process';
// eslint-disable-next-line no-restricted-imports
import events from 'node:events';
import path from 'node:path';

export const wranglerEntryPath = path.resolve(
  __dirname,
  '../node_modules/.bin/wrangler'
);

/**
 * Runs the command `wrangler dev` in a child process.
 *
 * Returns an object that gives you access to:
 *
 * - `ip` and `port` of the http-server hosting the pages project
 * - `stop()` function that will close down the server.
 */
export async function runWranglerDev(
  cwd: string,
  options: string[],
  env?: NodeJS.ProcessEnv
): Promise<{
  ip: string;
  port: number;
  stop: () => Promise<void>;
  getOutput: () => string;
  clearOutput: () => void;
}> {
  return runLongLivedWrangler(['dev', ...options], cwd, env);
}

async function runLongLivedWrangler(
  command: string[],
  cwd: string,
  env?: NodeJS.ProcessEnv
) {
  let settledReadyPromise = false;
  let resolveReadyPromise: (value: {ip: string; port: number}) => void;
  let rejectReadyPromise: (reason: unknown) => void;

  const ready = new Promise<{ip: string; port: number}>((resolve, reject) => {
    resolveReadyPromise = resolve;
    rejectReadyPromise = reject;
  });
  console.log(command);
  const wranglerProcess = fork(wranglerEntryPath, command, {
    stdio: [
      /* stdin */ 'ignore',
      /* stdout */ 'pipe',
      /* stderr */ 'pipe',
      'ipc',
    ],
    cwd,
    env: {...process.env, ...env, PWD: cwd},
  }).on('message', message => {
    if (settledReadyPromise) {
      return;
    }
    settledReadyPromise = true;
    clearTimeout(timeoutHandle);
    resolveReadyPromise(JSON.parse(message.toString()));
  });

  const chunks: Buffer[] = [];
  wranglerProcess.stdout?.on('data', chunk => {
    chunks.push(chunk);
  });
  wranglerProcess.stderr?.on('data', chunk => {
    chunks.push(chunk);
  });
  const getOutput = () => {
    return Buffer.concat(chunks).toString();
  };
  const clearOutput = () => {
    chunks.length = 0;
  };

  const timeoutHandle = setTimeout(() => {
    if (settledReadyPromise) {
      return;
    }
    settledReadyPromise = true;
    const separator = '='.repeat(80);
    const message = [
      'Timed out starting long-lived Wrangler:',
      separator,
      getOutput(),
      separator,
    ].join('\n');
    rejectReadyPromise(new Error(message));
  }, 50_000);

  async function stop() {
    const closePromise = events.once(wranglerProcess, 'close');
    wranglerProcess.kill('SIGTERM');
    const [code] = await closePromise;
    if (code) {
      throw new Error(`Exited with code ${code}`);
    }
  }

  const {ip, port} = await ready;
  return {ip, port, stop, getOutput, clearOutput};
}
