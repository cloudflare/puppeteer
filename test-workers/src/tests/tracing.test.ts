import {resolve} from 'path';

import {fetch} from 'undici';
import {afterAll, beforeAll, describe, it} from 'vitest';

import {runWranglerDev} from '../run-wrangler-long-lived.js';

describe('browser takes screenshot', () => {
  let ip: string,
    port: number,
    stop: (() => Promise<unknown>) | undefined,
    getOutput: () => string;

  // @ts-expect-error not sure why this is not working
  beforeAll(async () => {
    return ({ip, port, stop, getOutput} = await runWranglerDev(
      resolve(__dirname, '../..'),
      ['--remote']
    ));
  });

  afterAll(async () => {
    await stop?.();
    console.log(getOutput());
  });

  it('take sceenshot of example.com', async ({expect}) => {
    const response = await fetch(`http://${ip}:${port}/tracing`);
    expect(response.body).toMatchSnapshot();
  });
});
