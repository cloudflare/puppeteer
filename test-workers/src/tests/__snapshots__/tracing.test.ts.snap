// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`browser takes screenshot > take sceenshot of example.com 1`] = `
ReadableStream {
  Symbol(kType): "ReadableStream",
  Symbol(kState): {
    "controller": ReadableByteStreamController {
      Symbol(kType): "ReadableByteStreamController",
      Symbol(kState): {
        "autoAllocateChunkSize": undefined,
        "byobRequest": null,
        "cancelAlgorithm": [Function],
        "closeRequested": false,
        "highWaterMark": 0,
        "pendingPullIntos": [],
        "pullAgain": false,
        "pullAlgorithm": [Function],
        "pulling": false,
        "queue": [],
        "queueTotalSize": 0,
        "started": true,
        "stream": [Circular],
      },
    },
    "disturbed": false,
    "reader": undefined,
    "state": "readable",
    "storedError": undefined,
    "transfer": {
      "port1": undefined,
      "port2": undefined,
      "promise": undefined,
      "writable": undefined,
    },
  },
  Symbol(nodejs.webstream.isClosedPromise): {
    "promise": Promise {},
    "reject": [Function],
    "resolve": [Function],
  },
  Symbol(nodejs.webstream.controllerErrorFunction): [Function],
}
`;
